src/ray/pubsub/publisher.h

```diff
@@ -33,36 +33,68 @@ using SubscriberID = UniqueID;
 
 namespace pub_internal {
 
+class SubscriptionIndexInterface {
```

Comment:
These publisher changes could probably be pulled out into a separate PR, but they're pretty self-contained (makes the `SubscriptionIndex` follow the same pattern as the subscriber's `SubscriberChannel` to "properly" support templating.

Reply:
Makes sense. This is to put different templated index to the same dict right?